HW2/P2/parallel_vector.pyx

```diff
@@ -80,12 +79,15 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
    with nogil:
-       for r in range(repeat):
+       for r in prange(repeat):
            for idx in range(src.shape[0]):
                if counts[src[idx]] > 0:
```

Comment:
I think if you don't acquire the lock on counts[src[idx]] before the comparison to zero, you cannot guarantee that the counts elements will not fall to zero (or even negative) due to the actions of one or more threads by the time you have to subtract one from it below.